### PR TITLE
Introduced `model.isEmpty()` method. `DataController.get()` method can now trim empty data

### PR DESCRIPTION
**Pull request by f1ames**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced `model.isEmpty()` method. `DataController.get()` method can now trim empty data. Closes ckeditor/ckeditor5#401.

BREAKING CHANGE: `DataController.get()` method now returns an empty string by default when editor content is empty.

---

### Additional information

See ckeditor/ckeditor5#401. I have also created https://github.com/ckeditor/ckeditor5/tree/t/401 branch for easier testing.

There are a few related PRs which should be closed together with this one:

* https://github.com/ckeditor/ckeditor5-core/pull/161
* https://github.com/ckeditor/ckeditor5-block-quote/pull/32
* https://github.com/ckeditor/ckeditor5-enter/pull/61
* https://github.com/ckeditor/ckeditor5-paragraph/pull/41
* https://github.com/ckeditor/ckeditor5-table/pull/169
* https://github.com/ckeditor/ckeditor5-typing/pull/180

### CODE REVIEW

File: src/controller/datacontroller.js [hunk 1]
```diff
@@ -145,14 +154,21 @@ export default class DataController {
 	 *
 	 * @param {module:engine/model/element~Element|module:engine/model/documentfragment~DocumentFragment} modelElementOrFragment
 	 * Element whose content will be stringified.
+	 * @param {Boolean} [skipEmpty=false] Whether content considered empty should be skipped. The method
```
Comment (Reinmar) [inline src/controller/datacontroller.js]: I think that this logic should stay in `get()`. Less API changes is better.

File: src/controller/datacontroller.js [hunk 2]
```diff
@@ -115,10 +115,19 @@ export default class DataController {
...
-	 * @param {String} [rootName='main'] Root name.
+	 * @param {Object} [options]
+	 * @param {String} [options.rootName='main'] Root name.
+	 * @param {String} [options.trim='empty'] Whether returned data should be trimmed. This option is set to `empty` by default,
+	 * which means whenever editor content is considered empty, the empty string will be returned. To turn off trimming completely
+	 * use `none`. In such cases exact content will be returned (for example `<p>&nbsp;</p>` for empty editor).
 	 * @returns {String} Output data.
 	 */
-	get( rootName = 'main' ) {
+	get( options ) {
```
Comment (Reinmar) [inline src/controller/datacontroller.js]: This can be done with destructuring.

File: src/model/model.js [hunk 3]
```diff
@@ -472,6 +474,49 @@ export default class Model {
 		return false;
 	}
 
+	/**
+	 * Checks whether the given {@link module:engine/model/range~Range range} or
+	 * {@link module:engine/model/element~Element element} is considered empty.
+	 *
+	 * The range or element is considered non-empty if it contains any:
+	 * 		* EmptyElement
```
Comment (Reinmar) [inline src/model/model.js]: Can a model range contain a view empty element? ;)

File: src/model/model.js [hunk 4]
```diff
@@ -472,6 +474,49 @@ export default class Model {
...
+	 * 		* Non-plain `ContainerElement` (for example widget)
+	 * 		* Non-plain `AttributeElement` (for example comment)
+	 *
+	 * 	This method should be used to check if the element/range/editor contains any printable/meaningful content.
+	 * 	It is also the correct method to check if editor is empty.
+	 *
+	 * @param {module:engine/model/range~Range|module:engine/model/element~Element} rangeOrElement Range or element to check.
+	 * @returns {Boolean}
+	 */
+	isEmpty( rangeOrElement ) {
```
Comment (Reinmar) [inline src/model/model.js]: The relation between `isEmpty()` and `hasContent()` needs to be clarified. Both these methods are defined on `Model` but they have different semantics. I wonder if we shouldn't rename `hasContent()` or `isEmpty()` to somehow differentiate between these two.

cc @pjasiun

Comment (Reinmar) [reply src/model/model.js]: Let's merge `isEmpty()` into `hasContent()` (i.e. we will have a single method for that called `hasContent()`).

Right now there are 2 differences between them – check for markers and `/\S+/`. The former can be done automatically by `hasContent()`. The latter could be enabled by an `option.trimWhitespaces`.

Comment (f1ames) [reply src/model/model.js]: 👍

File: src/controller/datacontroller.js [hunk 5]
```diff
@@ -134,8 +140,10 @@ export default class DataController {
 			throw new CKEditorError( 'datacontroller-get-non-existent-root: Attempting to get data from a non-existing root.' );
 		}
 
+		const root = this.model.document.getRoot( rootName );
+
 		// Get model range.
-		return this.stringify( this.model.document.getRoot( rootName ) );
+		return trim === 'empty' && !this.model.hasContent( root, { trimWhitespaces: true } ) ? '' : this.stringify( root );
```
Comment (Reinmar) [inline src/controller/datacontroller.js]: This comment... ;)

**Comment by Espumisan** [comment]
I think the probelm in here is that there are no samples how to implement this in my project

**Pull request merged**